File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/jobs/movies/MovieCollectionJob.java
package com.battlelancer.seriesguide.jobs.movies;

import android.content.Context;
import android.support.annotation.NonNull;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.SgApp;
import com.battlelancer.seriesguide.jobs.episodes.JobAction;
import com.battlelancer.seriesguide.util.MovieTools;

public class MovieCollectionJob extends MovieJob {

    private final boolean isInCollection;

    public MovieCollectionJob(int movieTmdbId, boolean isInCollection) {
        super(isInCollection
                ? JobAction.MOVIE_COLLECTION_ADD : JobAction.MOVIE_COLLECTION_REMOVE, movieTmdbId);
        this.isInCollection = isInCollection;
    }

    @Override
    protected boolean applyDatabaseUpdate(Context context, int movieTmdbId) {
        MovieTools movieTools = SgApp.getServicesComponent(context).movieTools();
        return isInCollection
                ? movieTools.addToList(movieTmdbId, MovieTools.Lists.COLLECTION)
                : MovieTools.removeFromList(context, movieTmdbId, MovieTools.Lists.COLLECTION);
    }

    @NonNull
    @Override
    public String getConfirmationText(Context context) {
        return context.getString(isInCollection
                ? R.string.action_collection_add
                : R.string.action_collection_remove);
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/jobs/movies/MovieCollectionJob.kt
// SPDX-License-Identifier: GPL-3.0-or-later
// SPDX-FileCopyrightText: Copyright © 2020 Uwe Trottmann <uwe@uwetrottmann.com>

package com.battlelancer.seriesguide.jobs.movies

import android.content.Context
import com.battlelancer.seriesguide.SgApp.Companion.getServicesComponent
import com.battlelancer.seriesguide.jobs.episodes.JobAction
import com.battlelancer.seriesguide.movies.tools.MovieTools

/**
 * Adds or removes a movie from the collection.
 */
class MovieCollectionJob(
    movieTmdbId: Int,
    private val isInCollection: Boolean
) : MovieJob(
    if (isInCollection) JobAction.MOVIE_COLLECTION_ADD else JobAction.MOVIE_COLLECTION_REMOVE,
    movieTmdbId,
    0 /* Does not change plays. */
) {

    override suspend fun applyDatabaseUpdate(context: Context, movieTmdbId: Int): Boolean {
        val movieTools = getServicesComponent(context).movieTools()
        return if (isInCollection) {
            movieTools.addToList(movieTmdbId, MovieTools.Lists.COLLECTION)
        } else {
            movieTools.removeFromList(movieTmdbId, MovieTools.Lists.COLLECTION)
        }
    }
}